File: haskell/src/Y2015/D01.hs
{-# LANGUAGE LambdaCase #-}

module Y2015.D01
  ( d01
  ) where

import LibPuzzle

import Data.List
import Data.Maybe

d01 :: Computation
d01 = do
  moves <- fmap parseMove <$> readPuzzle 2015 1 "input"
  let floors      = scanl (+) 0 moves
      destFloor   = last floors
      basementIdx = fromMaybe (error "never reached the basement") (elemIndex (-1) floors)
  answers
    [ answer destFloor $ show destFloor ++ ": destination floor"
    , answer basementIdx $ show basementIdx ++ " move(s) to first visit the basement"
    ]

parseMove :: Char -> Int
parseMove = \case
  '(' -> 1
  ')' -> -1
  _   -> error "unexpected character"

File: rust/src/y2015/d01.rs
use failure::bail;

use crate::{Date, Day, Puzzle, Result};

const DATE: Date = Date::new(Day::D01, super::YEAR);
pub(super) const PUZZLE: Puzzle = Puzzle::new(DATE, solve);

#[allow(clippy::needless_pass_by_value)]
fn solve(input: String) -> Result {
    let mut moves = 0;
    let mut floor = 0;
    let mut visited_basement = false;
    for c in input.chars() {
        match c {
            '(' => floor += 1,
            ')' => floor -= 1,
            _ => bail!("unexpected character '{}'", c),
        }
        if !visited_basement {
            moves += 1;
            if floor == -1 {
                visited_basement = true;
            }
        }
    }
    answer!(floor, moves);
}
